core/src/main/java/org/apache/calcite/sql/SqlUtil.java

```diff
@@ -676,14 +680,16 @@ public static boolean matchRoutinesByParameterCount(
   }
 
   private static RelDataType bestMatch(List<SqlFunction> sqlFunctions, int i,
-      RelDataTypePrecedenceList precList) {
+      List<String> argNames, RelDataTypePrecedenceList precList) {
     RelDataType bestMatch = null;
     for (SqlFunction function : sqlFunctions) {
       List<RelDataType> paramTypes = function.getParamTypes();
       if (paramTypes == null) {
         continue;
       }
-      final RelDataType paramType = paramTypes.get(i);
+      final RelDataType paramType = argNames != null
+          ? paramTypes.get(function.getParamNames().indexOf(argNames.get(i)))
```

Comment:
Does all kinds of functions return a non-empty list if the method call have parameter names ? I don't think so, and how about `function.getParamNames().indexOf(argNames.get(i))` returns `-1` there ?

Reply:
@danny0405, good question!
This method is called after calling `filterRoutinesByParameterType()`, which will return `false` for the case when the incorrect name was specified in `argNames` list: https://github.com/apache/calcite/blob/1baee8524a8daa9c67f08f83080e269fc5938bc5/core/src/main/java/org/apache/calcite/sql/SqlUtil.java#L611
So at this point, we will have a list of functions, that have all parameter names from `argNames` list, if it is specified.

Reply:
Thanks for the explanation. The code is a mess. Maybe we should have a method named filterRoutinesByParameterName

Reply:
I have renamed `filterRoutinesByParameterType()` method and excracted logic connected with types and names permutation.